### PR TITLE
Add instructions on adding new Actions workflows

### PR DESCRIPTION
**Pull request by damccorm**
Once this is merged, I'll send an email to the dev list with similar instructions

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @Abacn @andreydevyatkin @tvalentyn @liferoad

File: .github/workflows/README.md
Comment (AnandInguva) [inline .github/workflows/README.md:23]: Does this checkout merges the PR with the master branch? If the user code is old, the test might fail with old errors that are already fixed on master

Comment (damccorm) [reply .github/workflows/README.md:23]: It does not, in that case the best path forward is for users to pull the latest contents from master.

In theory, you could do something a bit more fancy like what https://github.com/actions/checkout/issues/518 suggests, but that creates harder to follow/harder to get right branching depending on if you're doing the checkout as part of a PR or another trigger. My take is that its not worth it, especially since its easy to change down the line if this becomes a problem

File: .github/workflows/README.md
Comment (tvalentyn) [inline .github/workflows/README.md:95]: I took a look, looks like it UI allows select branches made on the main apache/beam repo. is it possible to point to a branch on my fork?

File: .github/workflows/README.md
Comment (tvalentyn) [inline .github/workflows/README.md:86]: If I were to add a test that will run in apache-beam-testing project, what kind of credentials would i need to run such test? guessing that uploading any service accounts would also be a security risk, so should we advise to request committers help to trigger test runs or mention that we can only test plumbing of the action but test execution will have to wait until the suite is merged?

File: .github/workflows/README.md
Comment (damccorm) [reply .github/workflows/README.md:95]: Yes, you just need to be on your fork's actions tab. I updated for clarity

File: .github/workflows/README.md
Comment (damccorm) [reply .github/workflows/README.md:86]: So we've basically decided to live with that risk for now; it is ok to provide access to limited service accounts when needed to adequately test pipelines. Typically, that only exposes using our compute as an attack vector since pipelines basically just need VMs or Dataflow to run, which is something we can respond to if it becomes an issue. The same set of risks exist on Jenkins today (since precommits can start arbitrary jobs using service account permissions on the boxes, and we decided to not take on the burden of fixing our full test model as part of this migration.

There's basically no way to allow users to run tests that use compute without either:
1) This risk
2) Requiring manual intervention for every precommit

File: .github/workflows/README.md
Comment (tvalentyn) [inline .github/workflows/README.md:62]: I think this is the most confusing part. just to make sure i got it right:

Is @ just a symbol in a string or it has a specific interpretation?

Does inclusion of  `${{ github.event.schedule }}` mean that we allow concurrent runs that were triggered at different times by a cron job (for example if a workflow takes 4 hrs to finish, but we run it every hour, we will have concurrent runs)

Does inclusion of `-${{ github.event.sender.login }}-`   mean that we will allow concurrent runs started by different users without cancelling them?

Comment (damccorm) [reply .github/workflows/README.md:62]: > Is @ just a symbol in a string or it has a specific interpretation?

It is just a symbol included by convention

> Does inclusion of ${{ github.event.schedule }} mean that we allow concurrent runs that were triggered at different times by a cron job (for example if a workflow takes 4 hrs to finish, but we run it every hour, we will have concurrent runs)

No, we should avoid having cron triggers that take longer to run than the schedule interval (they will be cancelled). This is partially just because I couldn't actually find a good differentiator in the actions payload between 2 scheduled runs, but I also think its a good idea (otherwise we're monopolizing more than 1 machine full time for a single test run).

> Does inclusion of -${{ github.event.sender.login }}- mean that we will allow concurrent runs started by different users without cancelling them?

Yes

Comment (damccorm) [reply .github/workflows/README.md:62]: >> Is @ just a symbol in a string or it has a specific interpretation?

> It is just a symbol included by convention

Here's an example of how these are rendered (this one doesn't include the schedule or username changes)

image

File: .github/workflows/README.md
Comment (tvalentyn) [reply .github/workflows/README.md:95]: and, I need to make sure to push beam's master to my fork's master too, right? I actually haven't been updating my fork's master for years : )

Comment (damccorm) [reply .github/workflows/README.md:95]: Oh interesting, I wrongly assumed that was part of most people's workflows! Yeah, you need to do that for this to work, I added a note on that as well

File: .github/workflows/README.md
Comment (tvalentyn) [reply .github/workflows/README.md:86]: Letting GH actions to run Dataflow pipelines seems reasonable. My question is about how to advise non-committers to test their actions workflows on Dataflow. Sounds like a committer would be able to fully test a suite using stored credentials for GH action, if they create a branch on Apache Repo. Non-committers wouldn't be able to create branches on Apache Repo, but they can test against branches created on their own fork; however their own fork would presumably not preconfigured GCP secrets. So likely, their gh action test will start, but dataflow pipeline will fail to launch, and that might be OK. We probably shouldn't advise customers to upload their own secrets to their own gh actions..

File: .github/workflows/README.md
Comment (tvalentyn) [reply .github/workflows/README.md:62]: Thanks. So then if we didn't include ${{ github.event.schedule }} in the group identifier,  we would have scheduled and unscheduled runs step on each other?

Comment (tvalentyn) [reply .github/workflows/README.md:62]: Also wondering, if I have Beam's actions workflows to my own fork, and there are cron schedules, will the actions continue to run indefinitely on my own fork? Sounds like it would be wasting actions' compute cycles... just wondering if there are some limits to the amount of available resources or whether we should maybe restrict workflows run only on apache/beam fork. This is unrelated to the PR.

File: .github/workflows/README.md
Comment (tvalentyn) [reply .github/workflows/README.md:86]: To make expectations clear we could say smth like: you would only be able to do X and Y but not Z, and if you really want to do Z you'd need Beam's committer help or wait until PR is merged or do XYZ (configure your own credentials, which I am hesitant to recommend).

**Review by tvalentyn — APPROVED** [review]
Thanks a lot!

File: .github/workflows/README.md
Comment (damccorm) [reply .github/workflows/README.md:62]: > So then if we didn't include ${{ github.event.schedule }} in the group identifier, we would have scheduled and unscheduled runs step on each other?

Yeah

> if I have Beam's actions workflows to my own fork, and there are cron schedules, will the actions continue to run indefinitely on my own fork?

Yeah it will. On the individual level, people can disable actions on their forks in the settings. But more broadly its probably a good idea to restrict this with a condition. I'll follow up with the team

**Pull request merged**
